### PR TITLE
Feature/bdx

### PR DESCRIPTION
**Pull request by meletios**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -30,7 +30,7 @@
             <dependency>
                 <groupId>org.springframework.cloud</groupId>
                 <artifactId>spring-cloud-dependencies</artifactId>
-                <version>Finchley.SR1</version>
+                <version>Greenwich.RELEASE</version>
```
Comment (palmfjord) [inline pom.xml]: should we really do this? :/

File: src/main/java/com/hedvig/claims/query/ClaimReportHistoryRepository.java [hunk 2]
```diff
@@ -1,10 +1,16 @@
...
 
+import java.time.Instant;
+import java.time.LocalDate;
 import java.util.List;
 import java.util.UUID;
 
 public interface ClaimReportHistoryRepository extends JpaRepository<ClaimReportHistoryEntity, UUID> {
   List<ClaimReportHistoryEntity> findByClaimId(String claimId);
+
+  @Query("from ClaimReportHistoryEntity where timeOfKnowledge >= :from AND timeOfKnowledge <= :to")
```
Comment (palmfjord) [inline src/main/java/com/hedvig/claims/query/ClaimReportHistoryRepository.java]: no caps? :c

Comment (palmfjord) [reply src/main/java/com/hedvig/claims/query/ClaimReportHistoryRepository.java]: technically _some_ caps? :cc

File: src/main/java/com/hedvig/claims/services/ReportGenerationService.java [hunk 3]
```diff
@@ -15,4 +16,6 @@
   void replay(YearMonth yearMonth);
 
   List<MiReportClaimHistoryDTO> generateMiReport(YearMonth until);
+
+  List<BDXReportClaimHistoryDTO> generateBDXReport(YearMonth month);
```
Comment (palmfjord) [inline src/main/java/com/hedvig/claims/services/ReportGenerationService.java]: i think we should be consistent in how we treat acronyms (`generateMiReport` vs `generateBDXReport`).

File: src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java [hunk 4]
```diff
@@ -50,14 +54,8 @@ public YearMonth getReportPeriod() {
...
 
-    final List<ClaimEntity> testClaims = claimsRepository.findByType(TEST);
-
-    final List<String> excludedClaimIds = testClaims.stream()
-      .map(claimEntity -> claimEntity.id)
-      .collect(Collectors.toList());
-
     final Map<String, List<ClaimReportHistoryEntity>> claimHistoryEntities = claimReportHistoryRepository.findAll().stream()
-      .filter(historyEntity -> !excludedClaimIds.contains(historyEntity.getClaimId()))
+      .filter(historyEntity -> !fetchCurrentClaimsByType().contains(historyEntity.getClaimId()))
```
Comment (palmfjord) [inline src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java]: do we need to do this at every history event? Feels inefficient

File: src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java [hunk 5]
```diff
@@ -99,18 +97,8 @@ public void replay(YearMonth yearMonth) {
...
-
-    List<String> excludedNotCoveredClaims = notCoveredClaims.stream().map(claimEntity -> claimEntity.id).collect(Collectors.toList());
-
-    List<String> excludedClaimIds = testClaims.stream().map(claimEntity -> claimEntity.id).collect(Collectors.toList());
-
-    excludedClaimIds.addAll(excludedNotCoveredClaims);
-
     return this.claimReportHistoryRepository.findAll().stream()
-      .filter(historyEntity -> !excludedClaimIds.contains(historyEntity.getClaimId()))
+      .filter(historyEntity -> !fetchCurrentClaimsByType().contains(historyEntity.getClaimId()))
```
Comment (palmfjord) [inline src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java]: should probably be done outside the stream also

File: src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java [hunk 6]
```diff
@@ -123,4 +111,39 @@ public void replay(YearMonth yearMonth) {
       .map(MiReportClaimHistoryDTO::from)
       .collect(Collectors.toList());
   }
+
+  public List<BDXReportClaimHistoryDTO> generateBDXReport(YearMonth month) {
+    Map<String, List<ClaimReportHistoryEntity>> claimsByClaimId = claimReportHistoryRepository
+      .findByTimeOfKnowledge(month.atDay(1).atTime(0, 0, 0, 0)
```
Comment (palmfjord) [inline src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java]: i think we could use `atStartOfDay` no?

Comment (palmfjord) [reply src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java]: also could we fix indentation/formatting/move stuff to variables here? I don't really understand what belongs to what without opening intellij

File: src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java [hunk 7]
```diff
@@ -123,4 +111,39 @@ public void replay(YearMonth yearMonth) {
...
+        .atZone(ZoneId.of(EUROPE_STOCKHOLM))
+        .toInstant(), month.atEndOfMonth()
+      .atTime(23, 59, 59, 999_999_999)
+      .atZone(ZoneId.of(EUROPE_STOCKHOLM))
+      .toInstant())
+      .stream()
+      .filter(claimReportHistoryEntity -> !fetchCurrentClaimsByType().contains(claimReportHistoryEntity.getClaimId()))
```
Comment (palmfjord) [inline src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java]: outside stream :D

File: src/main/java/com/hedvig/claims/web/ReportController.java [hunk 8]
```diff
@@ -15,9 +16,13 @@
 @RequestMapping("/report")
 public class ReportController {
 
-  @Autowired
   private ReportGenerationService reportGenerationService;
 
+  @Autowired
```
Comment (palmfjord) [inline src/main/java/com/hedvig/claims/web/ReportController.java]: much better hehe 😬

File: src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java [hunk 9]
```diff
@@ -123,4 +111,39 @@ public void replay(YearMonth yearMonth) {
...
+      .collect(Collectors.groupingBy(ClaimReportHistoryEntity::getClaimId));
+
+    Map<String, ClaimReportHistoryEntity> latestClaimsById = claimsByClaimId.entrySet().stream()
+      .map(
```
Comment (palmfjord) [inline src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java]: do we even need this? could we not just collect directly with 

```java
claimsByClaimId.entrySet().stream()
  .collect(toMap(
    entrySet -> entrySet.getKey(),
    entrySet -> entrySet.getValue()
      .max(Comparator.comparing(ClaimReportHistoryEntity::getTimeOfKnowledge))
            .get()
  ))
```

File: src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java [hunk 10]
```diff
@@ -98,19 +95,9 @@ public void replay(YearMonth yearMonth) {
...
-    List<String> excludedNotCoveredClaims = notCoveredClaims.stream().map(claimEntity -> claimEntity.id).collect(Collectors.toList());
-
-    List<String> excludedClaimIds = testClaims.stream().map(claimEntity -> claimEntity.id).collect(Collectors.toList());
-
-    excludedClaimIds.addAll(excludedNotCoveredClaims);
-
+  public List<ReportClaimHistoryDTO> generateClaimsReport(YearMonth until) {
     return this.claimReportHistoryRepository.findAll().stream()
-      .filter(historyEntity -> !excludedClaimIds.contains(historyEntity.getClaimId()))
+      .filter(historyEntity -> !fetchCurrentTestClaims().contains(historyEntity.getClaimId()))
```
Comment (palmfjord) [inline src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java]: maybe we shouldn't do this every time

Comment (meletios) [reply src/main/java/com/hedvig/claims/services/ReportGenerationServiceImpl.java]: Correcto!

**Pull request merged**